backends/tc/ebpfCodeGen.cpp

```diff
@@ -0,0 +1,1481 @@
+    builder->emitIndent();
+    builder->appendFormat("u32 %s = 0;", zeroKey.c_str());
+    builder->newline();
+    builder->emitIndent();
+    builder->appendFormat("u32 %s = 1;", oneKey.c_str());
+    builder->newline();
+    builder->emitIndent();
+    builder->appendFormat("unsigned char %s;", byteVar.c_str());
+    builder->newline();
+
+    builder->emitIndent();
+    builder->appendFormat("u32 %s = ", lengthVar.c_str());
+    emitPacketLength(builder);
+    builder->endOfStatement(true);
+
+    if (shouldEmitTimestamp()) {
+        builder->emitIndent();
+        builder->appendFormat("u64 %s = ", timestampVar.c_str());
+        emitTimestamp(builder);
+        builder->endOfStatement(true);
+    }
+
+    builder->emitIndent();
+    builder->appendFormat("u32 %s = %s;", inputPortVar.c_str(), ifindexVar.c_str());
+    builder->newline();
+}
+
+// =====================EBPFPnaParser=============================
+void EBPFPnaParser::emit(EBPF::CodeBuilder *builder) {
+    builder->emitIndent();
+    builder->blockStart();
+    EBPF::EBPFParser::emit(builder);
+    builder->blockEnd(true);
+}
+
+// =====================EBPFTablePNA=============================
+void EBPFTablePNA::emitKeyType(EBPF::CodeBuilder *builder) {
+    builder->emitIndent();
+    builder->appendFormat("struct __attribute__((__packed__)) %s ", keyTypeName.c_str());
+    builder->blockStart();
+
+    EBPF::CodeGenInspector commentGen(program->refMap, program->typeMap);
+    commentGen.setBuilder(builder);
+
+    unsigned int structAlignment = 4;  // 4 by default
+
+    builder->emitIndent();
+    builder->appendLine("u32 keysz;");
+    builder->emitIndent();
+    builder->appendLine("u32 maskid;");
+
+    if (keyGenerator != nullptr) {
+        for (auto c : keyGenerator->keyElements) {
+            auto mtdecl = program->refMap->getDeclaration(c->matchType->path, true);
+            auto matchType = mtdecl->getNode()->to<IR::Declaration_ID>();
+
+            if (!isMatchTypeSupported(matchType)) {
+                ::error(ErrorType::ERR_UNSUPPORTED, "Match of type %1% not supported",
+                        c->matchType);
+            }
+
+            auto ebpfType = ::get(keyTypes, c);
+            cstring fieldName = ::get(keyFieldNames, c);
+
+            if (ebpfType->is<EBPF::EBPFScalarType>() &&
+                ebpfType->to<EBPF::EBPFScalarType>()->alignment() > structAlignment) {
+                structAlignment = 8;
+            }
+
+            builder->emitIndent();
+            ebpfType->declare(builder, fieldName, false);
+
+            builder->append("; /* ");
+            c->expression->apply(commentGen);
+            builder->append(" */");
+            builder->newline();
+        }
+    }
+
+    // Add dummy key if P4 table define table with empty key. This due to that hash map
+    // cannot have zero-length key. See function htab_map_alloc_check in kernel/bpf/hashtab.c
+    // located in Linux kernel repository
+    if (keyFieldNames.empty()) {
```

Comment:
This can be removed - the limitation of an empty P4 key is related to the limitation of BPF hash maps. I assume that the P4TC subsystem will handle this properly.

Reply:
Removed